### ISSUE TITLE
timeout in //tensorflow/python/keras:data_utils_test when building from source

### ISSUE THREAD

**Issue by calid** [tensorflow/tensorflow#22804]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
```
$ uname -a
Linux precision 4.15.0-34-generic #37~16.04.1-Ubuntu SMP Tue Aug 28 10:44:06 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/lsb-release
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=18.2
DISTRIB_CODENAME=sonya
DISTRIB_DESCRIPTION="Linux Mint 18.2 Sonya"
```

CPU: Intel(R) Xeon(R) CPU E3-1505M v6 @ 3.00GHz
Mem: 32GB
Storage: 500GB

- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
No

- **TensorFlow installed from (source or binary)**:
Building from source

- **TensorFlow version (use command below)**:

    Issue occurs on both the latest 1.11.0 release branch and on latest master (573985c currently)

- **Python version**:
```
$ /usr/bin/python3 -V
Python 3.5.2
```

- **Bazel version (if compiling from source)**:
```
$ bazel version
Build label: 0.17.2
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Sep 21 10:31:42 2018 (1537525902)
Build timestamp: 1537525902
Build timestamp as int: 1537525902
```

- **GCC/Compiler version (if compiling from source)**:
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.10' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10)
```

- **CUDA/cuDNN version**:
N/A

- **GPU model and memory**:
```
# Quadro M1200, 4GB (3D acceleration)
# Intel HD Graphics 630 (primary vga)
lspci | grep -iP '(vga|3d)'
00:02.0 VGA compatible controller: Intel Corporation Device 591d (rev 04)
01:00.0 3D controller: NVIDIA Corporation Device 13b6 (rev a2)
```

- **Exact command to reproduce**:

`bazel test -c opt --action_env PATH="$PATH" -- //tensorflow/... -//tensorflow/compiler/... -//tensorflow/contrib/lite/...`

or

`bazel test --test_verbose_timeout_warnings -c opt --action_env PATH="$PATH" -- //tensorflow/python/keras:data_utils_test`

(I added --action_env because I was having some issues getting bazel to find my installed python3)

### Describe the problem

This is the only test that fails for me when building from source.  It seems timeouts in this test have been a problem before, per a commit by @jlebar: https://github.com/tensorflow/tensorflow/commit/7a60167ba7718c23b0ed70d079bbb446f63a4fd9

### Source code / logs

```
$ bazel test --test_verbose_timeout_warnings -c opt --action_env PATH="$PATH" -- //tensorflow/python/keras:data_utils_test

...

TIMEOUT: //tensorflow/python/keras:data_utils_test (Summary)
      /home/calid/.cache/bazel/_bazel_calid/d95f42fa008125d605be7949a2399f3e/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/data_utils_test/test.log
[1 / 2] 1 / 1 tests, 1 failed;  1 action; last test: .../python/keras:data_utils_test
Target //tensorflow/python/keras:data_utils_test up-to-date:
[2 / 2] 1 / 1 tests, 1 failed; no action; last  bazel-bin/tensorflow/python/keras/data_utils_test
[2 / 2] 1 / 1 tests, 1 failed; no action; lastINFO: Elapsed time: 915.423s, Critical Path: 915.01s
[2 / 2] 1 / 1 tests, 1 failed; no action; lastINFO: 1 process: 1 local.
[2 / 2] 1 / 1 tests, 1 failed; no action; lastINFO: Build completed, 1 test FAILED, 2 total actions
//tensorflow/python/keras:data_utils_test                               TIMEOUT in 915.0s
  /home/calid/.cache/bazel/_bazel_calid/d95f42fa008125d605be7949a2399f3e/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/data_utils_test/test.log

INFO: Build completed, 1 test FAILED, 2 total actions

$ cat /home/calid/.cache/bazel/_bazel_calid/d95f42fa008125d605be7949a2399f3e/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/data_utils_test/test.log
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/python/keras:data_utils_test
-----------------------------------------------------------------------------
Terminated
```

**Comment by calid** [comment]
After digging into this further the specific issue seems to be with using multiprocessing.  Every test that has `use_multiprocessing=True` hangs for me.  For example `test_ordered_enqueuer_processes`: https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/python/keras/utils/data_utils_test.py#L215-L224

Specifically, it hangs on `queue.get()` in `data_utils.py`: https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/python/keras/utils/data_utils.py#L605-L613

**Comment by calid** [comment]
I think I isolated the problem:  It looks like the `use_multiprocessing=True` branch of code is not actually handling the multiprocessing use case correctly, in that it isn't using the appropriate multiprocessing constructs (`multiprocessing.Queue`, `multiprocessing.Event`...).  

What's interesting is that in the `1.11.0` version `GeneratorEnqueuer` and `OrderedEnqueuer` handle multiprocessing differently, with `GeneratorEnqueuer` handling it correctly: 

`GeneratorEnqueuer.start (on 1.11.0)`:
https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/python/keras/utils/data_utils.py#L733-L743

vs `OrderedEnqueuer` which does not:

`OrderedEnqueuer.start (on 1.11.0)`:
https://github.com/tensorflow/tensorflow/blob/c19e29306ce1777456b2dbb3a14f511edf7883a8/tensorflow/python/keras/utils/data_utils.py#L549-L560

and sure enough on `1.11.0` only the OrderedEnqueuer tests hang, GeneratorEnqueuer tests do not.

On master it looks like there was an effort to coalesce the code and eliminate duplication by moving common logic into the SequenceEnqueuer base class.  Unfortunately it looks like the incorrect OrderedEnqueuer version of the code was used, and so now all the multiprocessing tests hang on master.

`SequenceEnqueuer.start (on master)`:
https://github.com/tensorflow/tensorflow/blob/573985cacba7ba36032fe3ac55f4ca9b52e08033/tensorflow/python/keras/utils/data_utils.py#L511-L521


I'm going to work on putting together a PR to use the correct multiprocessing constructs when `use_multiprocessing=True` (basically to just use the GeneratorEnqueuer version of the multiprocessing  logic).  Let me know if I'm on the right track with this or if I've misinterpreted things.

**Comment by calid** [comment]
@XuesongYang I've opened a PR to fix the issue: https://github.com/tensorflow/tensorflow/pull/23011.   If you have time want to give my branch a try and let me know if it fixes things for you too?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by calid** [comment]
Closing as the PR has moved to keras (which I'm still working on)
